### PR TITLE
porting removal of invoker, better decoupling

### PR DESCRIPTION
**Pull request by nayato**
Motivation:
Catching up on recent netty changes, simplifying data flow through channel pipeline

Modifications:
- Replaced IChannelHandlerInvoker with direct methods on IChannelHandlerContext implementation.
- Ported changes in AbstractChannel, DefaultChannelPipeline, etc. to reduce coupling through casting.

Result:
IChannelHandlerInvoker is removed, perf improved, it is easier to implement IChannel or IChannelPipeline from scratch.

### CODE REVIEW

**Comment by nayato** [comment]
@mtuchkov, @myagley PTAL.

File: src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs [hunk 1]
```diff
@@ -205,116 +245,432 @@ public bool HasAttribute<T>(AttributeKey<T> key)
...
             return this;
         }
 
-        public abstract IChannelHandler Handler { get; }
+        internal static void InvokeExceptionCaught(AbstractChannelHandlerContext next, Exception cause)
+        {
+            Contract.Requires(cause != null);
 
-        public IChannelHandlerContext FireChannelRead(object msg)
+            IEventExecutor executor = next.Executor;
```
Comment (nayato) [inline src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs]: nextExecutor here and elsewhere

File: src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs [hunk 2]
```diff
@@ -205,116 +245,432 @@ public bool HasAttribute<T>(AttributeKey<T> key)
...
+            if (executor.InEventLoop)
+            {
+                next.InvokeExceptionCaught(cause);
+            }
+            else
+            {
+                try
+                {
+                    executor.Execute( (c, e) => ((AbstractChannelHandlerContext)c).InvokeExceptionCaught((Exception)e), next, cause);
```
Comment (nayato) [inline src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs]: remove

File: src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs [hunk 3]
```diff
@@ -12,10 +12,18 @@ namespace DotNetty.Transport.Channels
...
     abstract class AbstractChannelHandlerContext : IChannelHandlerContext, IResourceLeakHint
     {
+        static readonly Action<object> InvokeChannelReadCompleteAction = ctx => ((AbstractChannelHandlerContext)ctx).InvokeChannelReadComplete();
+        static readonly Action<object> InvokeReadAction = ctx => ((AbstractChannelHandlerContext)ctx).InvokeRead();
+        static readonly Action<object> InvokeChannelWritabilityChangedAction = ctx => ((AbstractChannelHandlerContext)ctx).InvokeChannelWritabilityChanged();
+        static readonly Action<object> InvokeFlushAction = ctx => ((AbstractChannelHandlerContext)ctx).InvokeFlush();
+        static readonly Action<object, object> InvokeUserEventTriggeredAction = (ctx, evt) => ((AbstractChannelHandlerContext)ctx).InvokeUserEventTriggered(evt);
+        static readonly Action<object, object> InvokeChannelReadAction = (ctx, msg) => ((AbstractChannelHandlerContext)ctx).InvokeChannelRead(msg);
+
         internal const int MASK_HANDLER_ADDED = 1;
```
Comment (nayato) [inline src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs]: as these are now "original", let's switch to enum now.

**Comment by nayato** [comment]
>                                 "Can't invoke handlerRemoved() as the EventExecutor {} rejected it," +

capitalize

---
Refers to: src/DotNetty.Transport/Channels/DefaultChannelPipeline.cs:1264 in 4511b88. [](commit_id = 4511b88d5ad3946f061cc26927e8098e83144ac6, deletion_comment = False)

**Comment by nayato** [comment]
>                                 "Can't invoke handlerRemoved() as the EventExecutor {} rejected it," +

IE

---
Refers to: src/DotNetty.Transport/Channels/DefaultChannelPipeline.cs:1264 in 4511b88. [](commit_id = 4511b88d5ad3946f061cc26927e8098e83144ac6, deletion_comment = False)

File: src/DotNetty.Transport/Channels/ChannelHandlerInvokerUtil.cs [hunk 4]
```diff
@@ -1,247 +0,0 @@
...
-// Licensed under the MIT license. See LICENSE file in the project root for full license information.
-
-namespace DotNetty.Transport.Channels
-{
-    using System;
-    using System.Net;
-    using System.Threading.Tasks;
-    using DotNetty.Common.Utilities;
-
-    public static class ChannelHandlerInvokerUtil
```
Comment (nayato) [inline src/DotNetty.Transport/Channels/ChannelHandlerInvokerUtil.cs]: it's now folded into AbstractChannelHandlerContext

File: src/DotNetty.Transport/Channels/DefaultChannelHandlerInvoker.cs [hunk 5]
```diff
@@ -1,436 +0,0 @@
...
-    using System;
-    using System.Diagnostics.Contracts;
-    using System.Net;
-    using System.Threading.Tasks;
-    using DotNetty.Common;
-    using DotNetty.Common.Concurrency;
-    using DotNetty.Common.Internal;
-    using DotNetty.Common.Utilities;
-
-    public class DefaultChannelHandlerInvoker : IChannelHandlerInvoker
```
Comment (nayato) [inline src/DotNetty.Transport/Channels/DefaultChannelHandlerInvoker.cs]: It's now folded into AbstractChannelHandlerContext

File: src/DotNetty.Common/Concurrency/IEventExecutorGroup.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Common.Concurrency
+{
+    using System;
+    using System.Threading.Tasks;
+
+    public interface IEventExecutorGroup
```
Comment (nayato) [inline src/DotNetty.Common/Concurrency/IEventExecutorGroup.cs]: inherit IEventLoopGroup from this

File: src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs [hunk 7]
```diff
@@ -205,116 +247,432 @@ public bool HasAttribute<T>(AttributeKey<T> key)
...
+                    return this.Handler.ConnectAsync(this, remoteAddress, localAddress);
+                }
+                catch (Exception ex)
+                {
+                    return ComposeExceptionTask(ex);
+                }
+            }
+
+            return this.ConnectAsync(remoteAddress, localAddress);
         }
```
Comment (mtuchkov) [inline src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs]: just a bookmark for myself

File: src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs [hunk 8]
```diff
@@ -205,116 +247,432 @@ public bool HasAttribute<T>(AttributeKey<T> key)
...
+            {
+                try
+                {
+                    this.Handler.ExceptionCaught(this, cause);
+                }
+                catch (Exception t)
+                {
+                    if (DefaultChannelPipeline.Logger.WarnEnabled)
+                    {
+                        DefaultChannelPipeline.Logger.Warn("Failed to submit an ExceptionCaught() event.", t);
```
Comment (mtuchkov) [inline src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs]: Do we want to move it to resources eventually?

File: src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs [hunk 9]
```diff
@@ -324,24 +682,248 @@ public Task DisconnectAsync()
...
+            {
+                this.InvokeFlush0();
+            }
+            else
+            {
+                this.Flush();
+            }
         }
 
+        void InvokeFlush0()
```
Comment (mtuchkov) [inline src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs]: Is it a good method name?

Comment (nayato) [inline src/DotNetty.Transport/Channels/AbstractChannelHandlerContext.cs]: it is. Won't change for port-parity reasons.

**Comment by mtuchkov** [comment]
Looks good

**Pull request merged**